src/libappimage/libappimage.c

```diff
@@ -1890,12 +1888,36 @@ void appimage_create_thumbnail(const char *appimage_file_path, bool verbose) {
 
 }
 
-void appimage_extract_file_following_symlinks(const gchar* appimage_file_path, const char* file_path, const char* target_dir) {
-    appimage_handler handler = create_appimage_handler(appimage_file_path);
+void appimage_extract_file_following_symlinks(const gchar* appimage_file_path, const char* file_path,
+                                              const char* target_file_path) {
 
-    extract_appimage_file(&handler, file_path, target_dir);
+    struct extract_appimage_file_command_data data;
+    data.link = strdup(file_path);
+    data.destination = target_file_path;
 
-    // TODO: free handler?
+    bool looping = false;
+    GSList* visited_entries = NULL;
+
+    do {
+        visited_entries = g_slist_prepend(visited_entries, data.link);
+        data.path = data.link;
+        data.link = NULL;
+
+        appimage_handler handler = create_appimage_handler(appimage_file_path);
+        handler.traverse(&handler, extract_appimage_file_command, &data);
+
+        if (data.link) {
+            if (data.link && visited_entries &&
+                g_slist_find_custom(visited_entries, data.link, (GCompareFunc) strcmp))
+                looping = true;
+
+            g_remove(target_file_path);
+        }
+
+    } while (data.link && !looping);
```

Comment:
Is `data.link` a boolean? If not, please use a more explicit syntax.

Reply:
`data.link` is a `char*` this basically checks if it's  `NULL` or not.

Reply:
Then please do this explicitly, i.e. check against `NULL`. I _hate_ these kind of if clauses where you can only guess what's going on.

Reply:
Fixed

Reply:
Not really. There's many uses of `data.link`. Please fix all.

Reply:
Sorry. New changes were pushed